### PR TITLE
[rpc] createrawtransaction: Accept sorted outputs

### PR DESCRIPTION
**Pull request by maflcko**
The second parameter of the `createrawtransaction` is a dictionary of the outputs. This comes with at least two drawbacks:

* In case of duplicate keys, either of them might silently disappear, with no user feedback at all. A user needs to make other mistakes, but this could eventually lead to abnormal tx fees.
* A dictionary does not guarantee that keys are sorted. Again, a user needs to keep this in mind, as it could eventually lead to excessive tx fees.

Even though my scenario of loss-of-funds is unlikely to happen, I see it as a inconvenience that should be fixed.

### CODE REVIEW

**Comment by sipa** [comment]
Vague concept ACK that accepting a list instead of a dick should be supported. I have little opinion on how to do that in a backward compatible way.

**Comment by jonasschnelli** [comment]
Concept ACK.
I think the backward compatibility is solved in a good way... not sure if it should be mentioned somewhere (at least in some comments)

**Comment by promag** [comment]
Should use deprecated flag to enable backward compatibility?

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -349,14 +355,25 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
...
+        true);
     if (request.params[0].isNull() || request.params[1].isNull())
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, arguments 1 and 2 must be non-null");
 
     UniValue inputs = request.params[0].get_array();
-    UniValue sendTo = request.params[1].get_obj();
+    const bool outputs_is_array = request.params[1].isArray();
+    UniValue outputs = outputs_is_array ?
+                           request.params[1].get_array() :
+                           request.params[1].get_obj();
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: This will raise `JSON value is not an object as expected` which doesn't match with the new signature. I suggest to switch to `isObject()` above so that `get_array` throws. In that case, it is not necessary to call`get_obj()`:
```cpp
... = outputs_is_object ? request.params[1] : request.params[1].get_array();
```

**Review by promag — COMMENTED** [review]
> In case of duplicate keys, either of them might silently disappear, with no user feedback at all

Why? `Invalid parameter, duplicated address` is thrown.

> A dictionary does not guarantee that keys are sorted

Why does it matters?

Are these concerns because of the common lack support for duplicate keys in clients?

And since this is an utility call, weren't we moving enhancements to bitcoin-tx instead (although I'm ok with it)?

Maybe the missing piece is an `updaterawtransaction` call, where inputs/outputs can be added/removed/updated (just a thought)?

**Comment by promag** [comment]
@laanwj so just add a note that the parameter as object is deprecated?

**Comment by laanwj** [comment]
> TODO: Needs tests

Something that I guess needs to be tested explicitly now: duplicate address causes failure.

**Comment by MarcoFalke** [comment]
Note that this didn't break any functional test, so it should be compatible with the previous interface.

**Comment by jnewbery** [comment]
Concept ACK. I'll review once there are tests.

Have you seen luke-jr's change to allow RPC Type Checking against more than one type. It's in commit https://github.com/bitcoin/bitcoin/pull/11660/commits/8334875d8dcebab3841e42ab25e4e1d5247070a6 in #11660. Would it be useful here?

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp]: Can you provide code to reproduce that throw, please. When I tested it a day ago in the gui, it didn't throw.

**Comment by MarcoFalke** [comment]
> Why? Invalid parameter, duplicated address is thrown.

Not when the dict implementation eats duplicate keys. This is at least the case for python's dict and javascript's dict.

> Why does it matters?

The outputs structure is a vector, i.e. sorted in the bitcoin protocol, thus one could falsely assume that creating a raw transaction via `createrawtransaction` populates that vector in the same order as the outputs are typed. 

> Have you seen luke-jr's change to allow RPC Type Checking against more than one type. It's in commit 8334875 in #11660. Would it be useful here?

Yes, that could be useful. Though out of scope for this pull, imo. I am calling either `get_array` or `get_obj`, both of which throw on invalid type. So I am pretty sure I don't need to pull in further features.

C.f. https://dev.visucore.com/bitcoin/doxygen/univalue__get_8cpp_source.html#l00141

**Comment by laanwj** [comment]
> Not when the dict implementation eats duplicate keys. This is at least the case for python's dict and javascript's dict.

Normally whatever dict/object implementation is on the client-side will eat it.
To be able pass duplicate keys, the clients need to some emit an array as a JSON object.
That's absolutely against the JSON spec, just like expecting object attributes to be emitted in a certain order. To quote myself from IRC:

> \<wumpus\> FWIW in principle it is already possible to specify the order of outputs, but only if the client-side JSON library can emit ordered dictionaries. Python's can do this when you use Collections.OrderedDict() instead of a dictionary. The server-side (univalue) won't reorder dictionaries. But using an array is conforming to the spec instead of 'it happens to work'.

> Note that this didn't break any functional test, so it should be compatible with the previous interface.

It seems to break Travis, any idea why (or is this another fluke?).

**Comment by jnewbery** [comment]
> It seems to break Travis, any idea why (or is this another fluke?).

No, not a random Travis failure. This rpc unit test is broken:

```
BOOST_CHECK_THROW(CallRPC("createrawtransaction [] []"), std::runtime_error);
```
(src/test/rpc_tests.cpp, L55)

since `createrawtransaction` can now take an array as the second argument.

**Comment by MarcoFalke** [comment]
I will rebase and fixup all of this when the other createrawtransaction test are merged.

**Comment by instagibbs** [comment]
concept ACK

File: src/rpc/rawtransaction.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp]: @promag Heh, now I see what you mean. Makes sense.

Comment (promag) [reply src/rpc/rawtransaction.cpp]: Cool, I forgot to reply! :)

**Comment by promag** [comment]
BTW, echoing @laanwj https://github.com/bitcoin/bitcoin/pull/8457#issuecomment-237647103. I understand that here it's a different case though.

**Comment by MarcoFalke** [comment]
@promag Can you elaborate a bit more on that? It is a different case and in fact we do overload already, IIRC for some bool/integer parameters.

**Comment by promag** [comment]
Just wanted to link his concern regarding overloading. However here we want to fix a problem which happens to overload the RPC.

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -406,11 +423,21 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
...
-    std::vector<std::string> addrList = sendTo.getKeys();
-    for (const std::string& name_ : addrList) {
-
+    if (!outputs_is_obj) {
+        // Translate array of key-value pairs into dict
+        UniValue outputs_dict = UniValue(UniValue::VOBJ);
+        for (size_t i = 0; i < outputs.size(); ++i) {
+            const UniValue& output = outputs[i];
+            if (output.size() != 1) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter: Key-value pair must contain exactly one key");
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Current format: `Invalid parameter, key-value pair ...`.

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -406,11 +423,21 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
...
 
     std::set<CTxDestination> destinations;
-    std::vector<std::string> addrList = sendTo.getKeys();
-    for (const std::string& name_ : addrList) {
-
+    if (!outputs_is_obj) {
+        // Translate array of key-value pairs into dict
+        UniValue outputs_dict = UniValue(UniValue::VOBJ);
+        for (size_t i = 0; i < outputs.size(); ++i) {
+            const UniValue& output = outputs[i];
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Validate output is an object.

Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp]: Done. Also added a test for this

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -406,11 +423,24 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
...
+            if (!output.isObject()) {
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Use `RPCTypeCheckArgument` instead?

File: test/functional/rawtransactions.py [hunk 5]
```diff
@@ -263,7 +302,7 @@ def run_test(self):
         encrawtx = "01000000010000000000000072c1a6a246ae63f74f931e8365e15a089c68d61900000000000000000000ffffffff0100e1f505000000000000000000"
         decrawtx = self.nodes[0].decoderawtransaction(encrawtx, False) # decode as non-witness transaction
         assert_equal(decrawtx['vout'][0]['value'], Decimal('1.00000000'))
-        
+
```
Comment (promag) [inline test/functional/rawtransactions.py]: 👀

File: src/rpc/server.cpp [hunk 6]
```diff
@@ -67,10 +66,10 @@ void RPCTypeCheck(const UniValue& params,
     }
 }
 
-void RPCTypeCheckArgument(const UniValue& value, UniValue::VType typeExpected)
+void RPCTypeCheckArgument(const UniValue& value, const UniValueType& typeExpected)
```
Comment (promag) [inline src/rpc/server.cpp]: Move to different commit?

**Review by promag — COMMENTED** [review]
As a follow up, should the test framework be updated to use this variant?

File: src/rpc/rawtransaction.cpp (hunk 4 above)
Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp]: Would make the error message less specific

File: src/rpc/server.cpp (hunk 6 above)
Comment (MarcoFalke) [reply src/rpc/server.cpp]: Thx, done.

File: test/functional/rawtransactions.py (hunk 5 above)
Comment (MarcoFalke) [reply test/functional/rawtransactions.py]: Removing trailing white space should be fine now, since adding trailing white space is forbidden. So we converge to a single "code style".

**Comment by MarcoFalke** [comment]
> should the test framework be updated to use this variant?

Depends on the use case and your personal preference. In most cases there is no functional difference.

**Comment by promag** [comment]
> > should the test framework be updated to use this variant?
> 
> Depends on the use case and your personal preference. In most cases there is no functional difference.

Just saying, because IMO we should encourage this *variant*. Maybe tag for or add release note.

BTW, I think you can remove the TODO in the description.

utACK fa2e390.

File: src/rpc/rawtransaction.cpp (hunk 4 above)
Comment (promag) [reply src/rpc/rawtransaction.cpp]: Right.

**Comment by MarcoFalke** [comment]
> add release note

Will do immediately after merge, as I am not sure which branch this will end up on.

**Comment by promag** [comment]
Sorry already had the tag.

**Comment by MarcoFalke** [comment]
Rebased

**Comment by MarcoFalke** [comment]
Added release notes.

**Comment by luke-jr** [comment]
I don't see the use case... it might make sense for error checking, but this doesn't add a check for duplicate keys either...?

**Comment by MarcoFalke** [comment]
@luke-jr The check for duplicate keys is already there. See "Invalid parameter, duplicated address" in the functional tests.

File: src/test/rpc_tests.cpp [hunk 7]
```diff
@@ -52,7 +52,6 @@ BOOST_AUTO_TEST_CASE(rpc_rawparams)
     BOOST_CHECK_THROW(CallRPC("createrawtransaction"), std::runtime_error);
     BOOST_CHECK_THROW(CallRPC("createrawtransaction null null"), std::runtime_error);
     BOOST_CHECK_THROW(CallRPC("createrawtransaction not_array"), std::runtime_error);
-    BOOST_CHECK_THROW(CallRPC("createrawtransaction [] []"), std::runtime_error);
```
Comment (jnewbery) [inline src/test/rpc_tests.cpp]: Why not just change from `BOOST_CHECK_THROW()` to `BOOST_CHECK_NO_THROW()` instead of removing entirely? Passing two arrays is now permitted.

File: doc/release-notes.md [hunk 8]
```diff
@@ -61,7 +61,8 @@ RPC changes
 
 ### Low-level changes
 
-- The `fundrawtransaction` rpc will reject the previously deprecated `reserveChangeKey` option.
+- The `fundrawtransaction` RPC will reject the previously deprecated `reserveChangeKey` option.
+- The `createrawtransaction` RPC will now accept an array or dictionary (kept for compatibility) for the `outputs` parameter. This makes the result of the call deterministic.
```
Comment (jnewbery) [inline doc/release-notes.md]: Could use simpler language and be more explicit `makes the result of the call deterministic` -> `means the order of transaction outputs can be specified by the client` or similar

**Review by jnewbery — COMMENTED** [review]
utACK fa7a11a4c798eacc984762a379c206b0c7da8723

A few nits inline, but none should delay merge.

File: src/rpc/rawtransaction.cpp [hunk 9]
```diff
@@ -335,12 +336,17 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
             "       } \n"
             "       ,...\n"
             "     ]\n"
-            "2. \"outputs\"               (object, required) a json object with outputs\n"
+            "2. \"outputs\"               (array, required) a json array with outputs (key-value pairs)\n"
+            "   [\n"
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Nit, the indentation could be fixed.

Comment (promag) [reply src/rpc/rawtransaction.cpp]: The alignment too:
```
1. "inputs"                (array, required) A json array of json objects
     [
       {
         "txid":"id",    (string, required) The transaction id
         "vout":n,         (numeric, required) The output number
         "sequence":n      (numeric, optional) The sequence number
       }
       ,...
     ]
2. "outputs"               (array, required) a json array with outputs (key-value pairs)
   [
    {
      "address": x.xxx,    (obj, optional) A key-value pair. The key (string) is the bitcoin address, the value (float or string) is the amount in BTC
    },
    {
      "data": "hex"      (obj, optional) A key-value pair. The key must be "data", the value is hex encoded data
    }
    ,...                     More key-value pairs of the above form. For compatibility reasons, a dictionary, which holds the key-value pairs directly, is also
                             accepted as second parameter.
   ]
```

**Review by promag — COMMENTED** [review]
Is there a use case to create a raw transaction without outputs?

File: src/rpc/rawtransaction.cpp (hunk 9 above)
Comment (MarcoFalke) [reply src/rpc/rawtransaction.cpp]: Fixed alignment

File: src/test/rpc_tests.cpp (hunk 7 above)
Comment (MarcoFalke) [reply src/test/rpc_tests.cpp]: Thx. Moved the check to the functional tests

**Comment by MarcoFalke** [comment]
Force pushed the documentation changes.

File: src/test/rpc_tests.cpp (hunk 7 above)
Comment (jnewbery) [reply src/test/rpc_tests.cpp]: Why not change to `BOOST_CHECK_NO_THROW()`?

**Comment by jnewbery** [comment]
Tested ACK fa06dfce0f1f8e5e34f416c556590fa56b721788.

I'd still prefer to modify the unit test rather than dropping it, but this is ready for merge.

File: src/test/rpc_tests.cpp (hunk 7 above)
Comment (MarcoFalke) [reply src/test/rpc_tests.cpp]: @jnewbery I think we deprecated the existing rpc unit tests in favor of the functional tests which test the rpc interface. If you wanted reasonable coverage, you'd have to duplicate all the rpc unit tests based on the functional rpc tests. So yeah, I am not convinced that the unit tests add additional coverage...

**Comment by promag** [comment]
Other than the 1 space indentation nit:
```
2. "outputs"               (array, required) a json array with outputs (key-value pairs)
   [
    {
```
utACK fa06dfc.

File: src/test/rpc_tests.cpp (hunk 7 above)
Comment (jnewbery) [reply src/test/rpc_tests.cpp]: If you really think the unit tests are deprecated, they should be removed in a separate PR. It doesn't make sense to me to remove just this one case in this PR.

Anyway, as I said - I don't think this should hold up merge. I'm just puzzled that you'd chose a larger diff that reduces test coverage (the fact that this unit test failed in the first iteration of this PR tells me that it was testing something).

Comment (MarcoFalke) [reply src/test/rpc_tests.cpp]: Yes. Moving all those tests to the functional test suite makes sense.

File: src/rpc/rawtransaction.cpp [hunk 10]
```diff
@@ -352,14 +358,25 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
             + HelpExampleCli("createrawtransaction", "\"[{\\\"txid\\\":\\\"myid\\\",\\\"vout\\\":0}]\" \"{\\\"data\\\":\\\"00010203\\\"}\"")
```
Comment (conscott) [inline src/rpc/rawtransaction.cpp]: Need to add an example for the sorted format? 

Right now it only shows examples with the 'backwards compatible" format

**Comment by conscott** [comment]
Tested ACK fa06dfce0f1f8e5e34f416c556590fa56b721788 

Just left small inline comment

**Comment by MarcoFalke** [comment]
Added a commit to address the documentation nit. This seems ready for merge.
